### PR TITLE
Add token authentification to Nextcloud server

### PR DESCRIPTION
**Pull request by beccon4**
- add token authentification
- set default service info path
- 
Viele Grüße
Conrad Beckert

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +53,6 @@ object CheckCommand "nextcloud" {
 
 ## Changelog
 * 2019-05-08: initial version (koelle25)
-
+* 2023-04-07: add token athentification (beccon4)
```
Comment (koelle25) [inline README.md:123]: Can you add a newline after this?

File: check_nextcloud.php [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (koelle25) [inline check_nextcloud.php]: Please remove this newline at the beginning of the file

File: check_nextcloud.php [hunk 3]
```diff
@@ -22,36 +23,60 @@ function convert_filesize($bytes, $decimals = 2) {
...
 }
 
 $nchost = trim($options['H']);
-$ncuri = trim($options['U']);
+$ncuri = isset($options['U']) ? trim($options['U']) : "/ocs/v2.php/apps/serverinfo/api/v1/info";
 $ncssl = (isset($options['s']) && is_bool($options['s'])) ? $options['s'] : true;
-$ncuser = trim($options['u']);
-$ncpass = trim($options['p']);
-$ncurl = ($ncssl ? "https://" : "http://") . $ncuser . ":" . $ncpass . "@" . $nchost . $ncuri;
+$ncsit =  isset($options['T']) ? trim($options['T']) : "";
```
Comment (koelle25) [inline check_nextcloud.php]: Can you give this variable a better name? I don't know what 'sit' is standing for. I assume server-info-token? Better would probably be `$nctoken`.

File: check_nextcloud.php [hunk 4]
```diff
@@ -22,36 +23,60 @@ function convert_filesize($bytes, $decimals = 2) {
...
+$ncuser = isset($options['u']) ? trim($options['u']) : "";
+$ncpass = isset($options['p']) ? trim($options['p']) : "";
+
+$context=null;
+if($ncsit) {
+  $opts = array (
+    'http' => array (
+    'method' => 'GET',
```
Comment (koelle25) [inline check_nextcloud.php]: This line and the next one should be indented a level more.

**Review by koelle25 — CHANGES_REQUESTED** [review]
Thanks for your PR! Good addition. I only have a few changes requested, can you please include them and squash your commits afterwards?

File: README.md [hunk 5]
```diff
@@ -56,3 +57,4 @@ object CheckCommand "nextcloud" {
 * 2023-04-07: add token athentification (beccon4)
 ## Authors
 * [Kevin Köllmann](https://github.com/koelle25)
+* [Conrad Beckert](https://github.com/beccon4)
```
Comment (koelle25) [inline README.md:129]: I would appreciate a new section for contributors, like so:

```
## Contributors
* [Conrad Beckert](https://github.com/beccon4)
```

File: README.md [hunk 6]
```diff
@@ -8,7 +8,8 @@ This is a monitoring plugin for [icinga](https://www.icinga.com) to check the st
 ## Usage
 Try the plugin at the command line like this:
 ```
-/usr/bin/php ./check_nextcloud.php -H cloud.example.com -u apiuser -p apipassword
```
Comment (koelle25) [inline README.md:11]: A mention of the old user/password should still be there for backwards compatibility. Maybe like so:

> Try the plugin at the command line like this:
> ```
> /usr/bin/php ./check_nextcloud.php -H cloud.example.com -T access-token
> ```
> 
> For backwards compatibility the old user/password method is also possible. Just remember that the user used has to be in the admin group, so use appropriate measures (a strong password, HTTPS only, etc.):
> ```
> /usr/bin/php ./check_nextcloud.php -H cloud.example.com -u apiuser -p apipassword
> ```

File: check_nextcloud.php [hunk 7]
```diff
@@ -6,12 +6,13 @@
  * Monitoring plugin to check the status of nextcloud serverinfo app
  *
  * Copyright (c) 2019 Kevin Köllmann <mail@kevinkoellmann.de>
- *
+ * contribution: Conrad Beckert <kontakt@miradata.de>
```
Comment (koelle25) [inline check_nextcloud.php]: Your contribution will already be mentioned in the README and on the GitHub project itself, so it's not needed here in the actual script, too.

File: check_nextcloud.php [hunk 8]
```diff
@@ -6,12 +6,13 @@
...
  *
  * Usage: /usr/bin/php ./check_nextcloud.php -H cloud.example.com -u /ocs/v2.php/apps/serverinfo/api/v1/info
  *
  *
  * For more information visit https://github.com/koelle25/check_nextcloud
+ * For more information visit https://github.com/beccon4/check_nextcloud
```
Comment (koelle25) [inline check_nextcloud.php]: This line is also not needed (see above)

File: check_nextcloud.php [hunk 9]
```diff
@@ -22,21 +23,75 @@ function convert_filesize($bytes, $decimals = 2) {
...
+    else {
+        return 0;
+    }
+}       
+
 function show_helptext() {
   print "check_nextcloud.php - Monitoring plugin to check the status of nextcloud serverinfo app.\n
 You need to specify the following parameters:
-  -H:  hostname of the nextcloud instance, e.g. cloud.example.com
+  -H hostname: domain address of the nextcloud instance, e.g. cloud.example.com
```
Comment (koelle25) [inline check_nextcloud.php]: Removing the `:` is good I think, but please keep the spaces between the parameter and the description, it makes it much more readable.

File: check_nextcloud.php [hunk 10]
```diff
@@ -22,21 +23,75 @@ function convert_filesize($bytes, $decimals = 2) {
...
   -U:  uri of the nextcloud serverinfo api, you'll find it at https://cloud.example.com/settings/admin/serverinfo
-  -T:  authenticate using serverinfo token (either -T or -u and -p)
+  -T token: authenticate using serverinfo token (either -T or -u and -p)
```
Comment (koelle25) [inline check_nextcloud.php]: Same (add 2 more spaces to improve readability)

File: check_nextcloud.php [hunk 11]
```diff
@@ -22,21 +23,75 @@ function convert_filesize($bytes, $decimals = 2) {
...
+	files
+	shares
+	shares_user
+	shares_groups
+	shares_link
+	shares_fed
+	php_version
+	db_size
+  -c range: Critical Value (returns 2 CRITICAL if -P out of range)
+  -w range:  Warning Value  (returns 1 WARNING if -P out of range)
```
Comment (koelle25) [inline check_nextcloud.php]: Make it `<range>`, so it is more obvious it is a value to replace

File: check_nextcloud.php [hunk 12]
```diff
@@ -48,9 +103,12 @@ function show_helptext() {
 }
 
 $nchost = trim($options['H']);
-$ncuri = isset($options['U']) ? trim($options['U']) : "/ocs/v2.php/apps/serverinfo/api/v1/info";
-$ncssl = (isset($options['s']) && is_bool($options['s'])) ? $options['s'] : true;
-$ncsit =  isset($options['T']) ? trim($options['T']) : "";
+$ncuri  = isset($options['U']) ? trim($options['U']) : "/ocs/v2.php/apps/serverinfo/api/v1/info";
+$ncssl  = (isset($options['s']) && is_bool($options['s'])) ? $options['s'] : true;
+$ncsit  = isset($options['T'])? trim($options['T']) : "";
```
Comment (koelle25) [inline check_nextcloud.php]: `$nctoken` ?

File: check_nextcloud.php [hunk 13]
```diff
@@ -22,21 +23,75 @@ function convert_filesize($bytes, $decimals = 2) {
   return sprintf("%.{$decimals}f %s", $bytes / pow(1024, $factor), @$size[$factor]);
 }
 
+function check_range($pattern,$value) {
```
Comment (koelle25) [inline check_nextcloud.php]: Can you elaborate (add inline comments), what exactly this function and the if-clauses/preg_matches do? Please also add a DocBlock describing input parameters and return value.

File: check_nextcloud.php [hunk 14]
```diff
@@ -22,21 +23,75 @@ function convert_filesize($bytes, $decimals = 2) {
...
+
 function show_helptext() {
   print "check_nextcloud.php - Monitoring plugin to check the status of nextcloud serverinfo app.\n
 You need to specify the following parameters:
-  -H:  hostname of the nextcloud instance, e.g. cloud.example.com
+  -H hostname: domain address of the nextcloud instance, e.g. cloud.example.com
   -U:  uri of the nextcloud serverinfo api, you'll find it at https://cloud.example.com/settings/admin/serverinfo
-  -T:  authenticate using serverinfo token (either -T or -u and -p)
+  -T token: authenticate using serverinfo token (either -T or -u and -p)
+  -P:  Performance Data Parameter:
```
Comment (koelle25) [inline check_nextcloud.php]: Can you add examples to the README about using this parameter? Does it have to be given once with a comma separated list or multiple time or just `-P` and all the performance parameters are being gathered?

File: check_nextcloud.php (hunk 11 above)
Comment (koelle25) [inline check_nextcloud.php]: Also: Are the `-c <range>`  and `-w <range>` parameters used in conjunction with the `-P` parameter or are they standalone?

**Review by koelle25 — CHANGES_REQUESTED** [review]

File: check_nextcloud.php [hunk 15]
```diff
@@ -6,7 +6,7 @@
  * Monitoring plugin to check the status of nextcloud serverinfo app
  *
  * Copyright (c) 2019 Kevin Köllmann <mail@kevinkoellmann.de>
- * contribution: Conrad Beckert <kontakt@miradata.de>
+ * contribution and further development: Conrad Beckert <kontakt@miradata.de>
```
Comment (koelle25) [inline check_nextcloud.php]: I'm still active on this project. Please remove this line.

File: check_nextcloud.php [hunk 16]
```diff
@@ -182,46 +182,46 @@ function show_helptext() {
 // print output for icinga
 if ($statuscode == 200) {
     $perf_data="";
-    $status_message = sprintf("Nextcloud Version: %s ",$nc_version);
+    $status_message = ""; # sprintf("Nextcloud Version: %s ",$nc_version);
```
Comment (koelle25) [inline check_nextcloud.php:214]: I think you can just remove the commented out part.

**Comment by Zottel92** [comment]
I would also be interested in the Token argument.
@beccon4 are you still working on the requested changes?
Or should I create a fork myself?

Best regards and thank you two for your work.

File: check_nextcloud.php (hunk 9 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 10 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 11 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 3 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 4 above)
Comment (beccon4) [reply check_nextcloud.php]: OK

File: README.md (hunk 5 above)
Comment (beccon4) [reply README.md:129]: ok

File: check_nextcloud.php (hunk 15 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 8 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

File: check_nextcloud.php (hunk 11 above)
Comment (beccon4) [reply check_nextcloud.php]: ok

**Comment by koelle25** [comment]
@beccon4 This PR is getting a bit out of hand 😅 May I ask you to create new ones with feature-specific changes/commits?

**Pull request closed without merging**